### ISSUE TITLE
Extreme memory usage and no way of freeing resources

### ISSUE THREAD

**Issue by ozontm** [stefan-baumann/ManiaPlanetSharp#39]
Through runtime of the app the memory usage in my app increases with every file that is parsed and read using the MapMetadataProvider. GC is not freeing up nearly as many resources as expected. After the MapMetadataProvider is not required anymore the resources that were created upon reading single properties seem to remain in memory.

In the branch MemoryTest I have put an example of what I'm encountering, including a map from TMX that raises the memory usage to over 50 GB (in comparison, usually it raises by 500 MB when loading a map). That seems to be an issue as well, as there are quite a few maps in our db that cause this.

![image](https://github.com/user-attachments/assets/8cceae1d-5d09-48c6-bd36-8da5b21341d2)

Is there a way to free up these resources?

**Comment by stefan-baumann** [comment]
Is this a new issue with the recent PR merges or new game updates? I don't remember this ever being an issue?

**Comment by ozontm** [comment]
> Is this a new issue with the recent PR merges or new game updates? I don't remember this ever being an issue?

No, this also occured in 9e3dfadc (before any commit by Tomek)

**Comment by stefan-baumann** [comment]
Hm, that definitely is troubling. I assume the pattern matching erroneously starts trying to parse some chunk and then allocates a lot of memory for some array or so. We've previously looked at this for strings and limited the maximum allocation lengths - maybe doing the same for arrays could help? A length check could, e.g., be added at https://github.com/stefan-baumann/ManiaPlanetSharp/blob/master/src/ManiaPlanetSharp/GameBox/Parsing/ParserGeneration/ParserCodeGenerator.cs#L174, choosing some value such as 100k or 1M that shouldn't cause any practical issues but prevent these allocations. I think it should just be enough to throw an exception if it's beyond the limit. Can you maybe give this a try yourself? I do not currently have the setup ready to build and test MP# unfortunately.

**Comment by ozontm** [comment]
```builder.AppendLine($"result.{field.Property.Name} = new {field.Property.PropertyType.GetElementType()}[{lengthSource}];");
builder.AppendLine($"if (result.{field.Property.Name}.Length > 1000000)");
builder.AppendLine("{");
builder.Indent();
builder.AppendLine($"throw new NotSupportedException(\"array length larger than 100k\");");
builder.UnIndent();
builder.AppendLine("}");```

Adding this has unfortunately not resolved the memory increase. However, some NotSupportedExceptions were thrown.

Note that the memory bump happens exactly at the moment where I request `provider.Name` in the screenshot above.

**Comment by stefan-baumann** [comment]
Hmm, this sounds really weird.
I guess adding this kind of sanity check could help prevent some memory usage in practice, but it's weird that it isn't the actual issue. I would assume that the header-based sources are not available and the method then falls back on the body, parsing it in the process, which then causes the problems. But if it's neither an array nor a string (those are already limited in length), I don't know what the cause could be. I previously introduced a LongString type that does not perform these sanity checks (back when there were issues with un-parseable maps due to that sanity check) that is used at https://github.com/stefan-baumann/ManiaPlanetSharp/blob/2200b85e20af6cbe58d7303e76d87bbefa0c45c0/src/ManiaPlanetSharp/GameBox/Parsing/Chunks/Ghost/GhostChunk.cs#L34, which could still be causing it. Other than that, the only thing I can think of is it being one of the `CustomParserMethod`s allocating some really large amounts of memory due to a similar problem.

Unfortunately, I do not have the setup to try this myself anymore, the best I can offer currently is some guidance.

**Comment by ozontm** [comment]
Longstring does not seem to be the issue (checked the calls in the generated chunkparsers.cs).

For now, I have remedied the issue by manually calling `GC.collect()` in my app. The spike (> 50 GB) is still there, but after parsing at least my memory is back to normal. Would go back into analysis at a later point.

**Comment by tomek0055** [comment]
I investigated this problem myself when checking why old maps parsed by MP# have blocks with incorrect identifiers. I analyzed this problem by logging memory usage before and after chunk parsing.

👉 Chunks with legacy ids fail to look up the skippable flag. If an array is the first member in the chunk, an array of 0x534b4950 elements is created, which allocates a lot of memory.
👉 I noticed a huge memory spike while parsing the 0x03029002 chunk. This chunk is implemented incorrectly. It turns out that this chunk is incorrectly marked as a non-skipable chunk and has only one field.

I have noticed that memory remains allocated when an exception occurs. Probably resources generated by an incorrectly parsed chunk are not automatically freed and remain in memory. I implemented my changes locally and noticed no more memory spikes on both old and new maps with the 0x03029002 chunk.

Pull request with changes has been created here: #40
